python/lsst/daf/butler/registry/attributes.py

```diff
@@ -0,0 +1,114 @@
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+from __future__ import annotations
+"""The default concrete implementation of the class that manages
+attributes for `Registry`.
+"""
+
+__all__ = ["DefaultButlerAttributeManager"]
+
+from typing import (
+    ClassVar,
+    Iterable,
+    Optional,
+    Tuple,
+)
+
+import sqlalchemy
+
+from ..core.ddl import TableSpec, FieldSpec
+from .interfaces import (
+    Database,
+    ButlerAttributeExistsError,
+    ButlerAttributeManager,
+    StaticTablesContext
+)
+
+
+class DefaultButlerAttributeManager(ButlerAttributeManager):
+    """An implementation of `ButlerAttributeManager` that stores attributes
+    in a database table.
+
+    Parameters
+    ----------
+    db : `Database`
+        Database engine interface for the namespace in which this table lives.
+    table : `sqlalchemy.schema.Table`
+        SQLAlchemy representation of the table that stores attributes.
+    """
+    def __init__(self, db: Database, table: sqlalchemy.schema.Table):
+        self._db = db
+        self._table = table
+
+    _META_TABLE_NAME: ClassVar[str] = "butler_attributes"
+
+    _META_TABLE_SPEC: ClassVar[TableSpec] = TableSpec(
+        fields=[
+            FieldSpec("name", dtype=sqlalchemy.String, length=1024, primaryKey=True),
+            FieldSpec("value", dtype=sqlalchemy.String, length=65535, nullable=False),
+        ],
+    )
+
+    @classmethod
+    def initialize(cls, db: Database, context: StaticTablesContext) -> ButlerAttributeManager:
+        # Docstring inherited from ButlerAttributeManager.
+        table = context.addTable(cls._META_TABLE_NAME, cls._META_TABLE_SPEC)
+        return cls(db=db, table=table)
+
+    def get(self, name: str, default: Optional[str] = None) -> Optional[str]:
+        # Docstring inherited from ButlerAttributeManager.
+        sql = sqlalchemy.sql.select([self._table.columns.value]).where(
+            self._table.columns.name == name
+        )
+        row = self._db.query(sql).fetchone()
+        if row is not None:
+            return row[0]
+        return default
+
+    def set(self, name: str, value: str, *, force: bool = False) -> None:
+        # Docstring inherited from ButlerAttributeManager.
+        if force:
+            self._db.replace(self._table, {
+                "name": name,
+                "value": value,
+            })
+        else:
+            try:
+                self._db.insert(self._table, {
+                    "name": name,
+                    "value": value,
+                })
+            except sqlalchemy.exc.IntegrityError as exc:
+                raise ButlerAttributeExistsError(f"attribute {name} already exists") from exc
```

Comment:
Maybe we won't be able to tell until we try to use this, but I suspect a version of this that uses `Database.sync` to implement "try to set k = v and raise if it is set to something else" might be more useful than the `insert` version, which will raise if k exists even if it's already set to the same value.

Reply:
The use case I was thinking about is "I know this attribute should not exist and I want to create it". I suspect that updating attributes will be a rare process under user control, so `force` option should cover that case.